File: services/workflow-automation-service/src/models/workflow.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use serde_json::Value;
use sqlx::FromRow;
use uuid::Uuid;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WorkflowBranchCondition {
    pub field: String,
    pub operator: String,
    pub value: Value,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WorkflowBranch {
    pub condition: WorkflowBranchCondition,
    pub next_step_id: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WorkflowStep {
    pub id: String,
    pub name: String,
    pub step_type: String,
    #[serde(default)]
    pub description: String,
    #[serde(default)]
    pub config: Value,
    #[serde(default)]
    pub next_step_id: Option<String>,
    #[serde(default)]
    pub branches: Vec<WorkflowBranch>,
}

#[derive(Debug, Clone, FromRow, Serialize, Deserialize)]
pub struct WorkflowDefinition {
    pub id: Uuid,
    pub name: String,
    pub description: String,
    pub owner_id: Uuid,
    pub status: String,
    pub trigger_type: String,
    pub trigger_config: Value,
    pub steps: Value,
    pub webhook_secret: Option<String>,
    pub next_run_at: Option<DateTime<Utc>>,
    pub last_triggered_at: Option<DateTime<Utc>>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

impl WorkflowDefinition {
    pub fn parsed_steps(&self) -> Result<Vec<WorkflowStep>, String> {
        serde_json::from_value(self.steps.clone()).map_err(|error| error.to_string())
    }
}

#[derive(Debug, Deserialize)]
pub struct CreateWorkflowRequest {
    pub name: String,
    pub description: Option<String>,
    pub status: Option<String>,
    pub trigger_type: String,
    #[serde(default)]
    pub trigger_config: Value,
    pub steps: Vec<WorkflowStep>,
}

#[derive(Debug, Deserialize)]
pub struct UpdateWorkflowRequest {
    pub name: Option<String>,
    pub description: Option<String>,
    pub status: Option<String>,
    pub trigger_type: Option<String>,
    pub trigger_config: Option<Value>,
    pub steps: Option<Vec<WorkflowStep>>,
}

#[derive(Debug, Deserialize)]
pub struct ListWorkflowsQuery {
    pub page: Option<i64>,
    pub per_page: Option<i64>,
    pub search: Option<String>,
    pub trigger_type: Option<String>,
    pub status: Option<String>,
}

File: openfoundry-go/services/workflow-automation-service/internal/models/workflow.go
// Package models holds the wire-format types for workflow-automation-service.
//
// Snake_case JSON tags + field order mirror the Rust serde encoding
// 1:1 so producers + UI clients can round-trip payloads against
// either runtime.
package models

import (
	"encoding/json"
	"time"

	"github.com/google/uuid"
)

// WorkflowBranchCondition mirrors models::workflow::WorkflowBranchCondition.
type WorkflowBranchCondition struct {
	Field    string          `json:"field"`
	Operator string          `json:"operator"`
	Value    json.RawMessage `json:"value"`
}

// WorkflowBranch mirrors models::workflow::WorkflowBranch.
type WorkflowBranch struct {
	Condition  WorkflowBranchCondition `json:"condition"`
	NextStepID string                  `json:"next_step_id"`
}

// WorkflowStep mirrors models::workflow::WorkflowStep.
type WorkflowStep struct {
	ID          string           `json:"id"`
	Name        string           `json:"name"`
	StepType    string           `json:"step_type"`
	Description string           `json:"description,omitempty"`
	Config      json.RawMessage  `json:"config,omitempty"`
	NextStepID  *string          `json:"next_step_id,omitempty"`
	Branches    []WorkflowBranch `json:"branches,omitempty"`
}

// WorkflowDefinition mirrors models::workflow::WorkflowDefinition. Rust
// uses sqlx FromRow; in Go we hydrate manually in repo helpers.
type WorkflowDefinition struct {
	ID              uuid.UUID       `json:"id"`
	Name            string          `json:"name"`
	Description     string          `json:"description"`
	OwnerID         uuid.UUID       `json:"owner_id"`
	Status          string          `json:"status"`
	TriggerType     string          `json:"trigger_type"`
	TriggerConfig   json.RawMessage `json:"trigger_config"`
	Steps           json.RawMessage `json:"steps"`
	WebhookSecret   *string         `json:"webhook_secret"`
	NextRunAt       *time.Time      `json:"next_run_at"`
	LastTriggeredAt *time.Time      `json:"last_triggered_at"`
	CreatedAt       time.Time       `json:"created_at"`
	UpdatedAt       time.Time       `json:"updated_at"`
}

// ParsedSteps mirrors WorkflowDefinition::parsed_steps.
func (w *WorkflowDefinition) ParsedSteps() ([]WorkflowStep, error) {
	if len(w.Steps) == 0 {
		return nil, nil
	}
	var out []WorkflowStep
	if err := json.Unmarshal(w.Steps, &out); err != nil {
		return nil, err
	}
	return out, nil
}

// CreateWorkflowRequest mirrors models::workflow::CreateWorkflowRequest.
type CreateWorkflowRequest struct {
	Name          string          `json:"name"`
	Description   *string         `json:"description,omitempty"`
	Status        *string         `json:"status,omitempty"`
	TriggerType   string          `json:"trigger_type"`
	TriggerConfig json.RawMessage `json:"trigger_config,omitempty"`
	Steps         []WorkflowStep  `json:"steps"`
}

// UpdateWorkflowRequest mirrors models::workflow::UpdateWorkflowRequest.
type UpdateWorkflowRequest struct {
	Name          *string          `json:"name,omitempty"`
	Description   *string          `json:"description,omitempty"`
	Status        *string          `json:"status,omitempty"`
	TriggerType   *string          `json:"trigger_type,omitempty"`
	TriggerConfig *json.RawMessage `json:"trigger_config,omitempty"`
	Steps         *[]WorkflowStep  `json:"steps,omitempty"`
}

// ListWorkflowsQuery mirrors models::workflow::ListWorkflowsQuery.
type ListWorkflowsQuery struct {
	Page        *int64
	PerPage     *int64
	Search      *string
	TriggerType *string
	Status      *string
}
